File: p.30/Haskell/solution.hs
module Main where

digitPowers :: [Int]
digitPowers = [ d^5 | d <- [0..9] ]

digits :: Int -> [Int]
digits 0 = [0]
digits n = reverse (go n [])
    where
        go 0 acc = acc
        go x acc =
            let (q, r) = x `divMod` 10
                in go q (r : acc)

isSumOfFifthPowers :: Int -> Bool
isSumOfFifthPowers n =
    let ds = digits n
        sumPowers = sum [ digitPowers !! d | d <- ds]
    in sumPowers == n

main :: IO ()
main = do
    let validNumbers = [ n | n <- [2..999999], isSumOfFifthPowers n]
    let answer = sum validNumbers
    print answer

File: p.30/Rust/solution.rs
fn build_digit_powers() -> [u32; 10] {
    let mut arr = [0; 10];
    for d in 0u32..10u32 {
        arr[d as usize] = d.pow(5);
    }
    arr
}

fn is_sum_of_fifth_powers(n: u32, digit_powers: &[u32; 10]) -> bool {
    let mut temp = n;
    let mut sum = 0;
    while temp > 0 {
        sum += digit_powers[(temp % 10) as usize];
        temp /= 10;
    }
    sum == n
}

fn main() {
    let digit_powers = build_digit_powers();

    let answer: u64 = (2..=999_999)
        .filter(|&x| is_sum_of_fifth_powers(x, &digit_powers))
        .map(|x| x as u64)
        .sum();

    println!("{}", answer);
}
